tests/nn_test.py

```diff
@@ -369,6 +371,38 @@ def testVarianceScalingError(self):
     ):
       initializer(rng, shape)
 
+  @parameterized.parameters([
+    nn.initializers.uniform(scale=np.sqrt(2), dtype=jnp.float64),
+    nn.initializers.normal(stddev=np.sqrt(2), dtype=jnp.float64),
+    nn.initializers.variance_scaling(np.sqrt(2), "fan_in", "uniform", dtype=jnp.float64),
+    nn.initializers.orthogonal(scale=np.sqrt(2), dtype=jnp.float64),
+    nn.initializers.truncated_normal(stddev=np.sqrt(2), dtype=jnp.float64),
+    nn.initializers.delta_orthogonal(scale=np.sqrt(2), dtype=jnp.float64),
+  ])
+  def testInitializerPreferUserRequestedDtype(self, initializer):
+    """Test that user requested dtype takes precedence."""
+    # See https://github.com/google/jax/issues/18267
+    if not config.enable_x64.value:
+      self.skipTest("requires x64 mode")
+    params = initializer(random.PRNGKey(0), (2, 2, 2), jnp.float32)
+    self.assertEqual(params.dtype, jnp.float32)
+
+  @parameterized.parameters([
+    nn.initializers.uniform(scale=np.sqrt(2), dtype=jnp.float32),
+    nn.initializers.normal(stddev=np.sqrt(2), dtype=jnp.float32),
+    nn.initializers.variance_scaling(np.sqrt(2), "fan_in", "uniform", dtype=jnp.float32),
+    nn.initializers.orthogonal(scale=np.sqrt(2), dtype=jnp.float32),
+    nn.initializers.truncated_normal(stddev=np.sqrt(2), dtype=jnp.float32),
+    nn.initializers.delta_orthogonal(scale=np.sqrt(2), dtype=jnp.float32),
+  ])
+  def testInitializerDtypeStrictDtypePromotion(self, initializer):
+    """Test that user requested dtype takes precedence."""
+    # See https://github.com/google/jax/issues/18267
+    if not config.enable_x64.value:
+      self.skipTest("requires x64 mode")
+    with self.assertRaises(dtypes.TypePromotionError):
+      initializer(random.PRNGKey(0), (2, 2, 2))
```

Comment:
Sorry to page back on this, but it seems strange that this would error, since the dtype is explicitly specified in the initializer instantiation. I think we need a different approach here.